lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -1099,6 +996,7 @@ bool mlir::triton::preProcessLoopAndGetSchedule(
     coarseSchedule.dump();
   });
 
+  LDBG("afterPrologue = " << *afterPrologue);
   scheduleRemainingToLastStage(forOp, coarseSchedule, afterPrologue, numStages);
```

Comment:
I think we need to consider what to do with the prologue/epilogue information. I was hoping that we can make the pipeliner agnostic of things like prologue and epilogue clusters, as those are (should be at least) schedule-dependent. Not sure what's the solution

Reply:
Why is it not possible to move them into the scheduler right now?

Reply:
Yeah I am not sure what the best solution here. My previous comment
```
We need to return afterPrologue from loopScheduling since it is needed in scheduleRemaining. The current rule between loopScheduling and SWP is that cluster with clusterId 0 will be afterPrologue (i.e clusterId for prologue in the case of 1D flattened persistent kernel will have negative clusterId from loopScheduling pass). This is an extra communication between the two passes which we should try to remove.
```
scheduleRemaining needs to know the first cluster excluding the prologue.

Reply:
> Why is it not possible to move them into the scheduler right now?

You mean moving scheduleRemainingToLastStage to the scheduler? It is currently run after scheduleDep, scheduleDistOne. If we want to keep scheduleDep inside SWP, we will probably need to keep scheduleRemainingToLastStage in SWP?
In scheduleRemainingToLastStage, an op will start from the cluster "afterPologue", if it uses the result of another op that is already in the schedule, it will be delayed till the cluster of the defining op.

Reply:
I'm starting to wonder if we want the scheduler to assign a stage and cluster to every op in the loop so that this kind of ambiguity can be removed. 
We would probably still want to first annotate the important ops first and then propagate but maybe that should be moved into the scheduler. This way the lowering would be completely independent of all that and only do lowering.
@pawelszczerbuk WDYT?

Reply:
The lowering in createAsyncOps happens before scheduleDep etc. If we move all scheduling out of SWP, we will need to assign schedule (stage, cluster) for the new ops created in createAsyncOps (that may be straight-forward?), and all ops in the loop body will have (stage, cluster) attributes after loopScheduling.
I guess that is fine?

Reply:
yes I think propagating the stage and clusters during lowering will be needed and makes sense